### PR TITLE
Add fieldset pattern

### LINKED ISSUE #234: Better, semantic simple_form patterns
_From @jrubenoff on February 10, 2016 17:42_

### Problem

Both of the labels here are `label` elements:

screen shot 2016-02-10 at 9 25 31 am

This isn't correct.

- It's not semantic. You can't have two labels for an input.

- It decreases usability. It's a label, so the browser gives you a pointer cursor on hover. But there's no `for` attribute, so nothing happens when you click on it.

### Solution

Add some better markup patterns to `simple_form`. A few examples:

screen shot 2016-02-10 at 9 30 53 am

screen shot 2016-02-10 at 9 37 19 am

screen shot 2016-02-10 at 9 39 56 am

screen shot 2016-02-10 at 9 34 02 am

_Copied from original issue: dobtco/screendoor-v2#3191_

### PR DESCRIPTION
**Pull request by jrubenoff**
Closes #234 

To-do:

- [ ] Convert date picker to a fieldset
  - [ ] Add `Choose a date` and `Choose a time` labels to date picker, and hide them with `@include hidden`

### CODE REVIEW

**Comment by ajb** [comment]
Is the "to do" for me?

Also, I would strongly recommend using either a generic utility class (`.hidden`), or a form-scoped utility class (`.form_label_hidden`) instead of scattering `@include hidden` everywhere... it'll save us a ton of bytes.

**Comment by jrubenoff** [comment]
The to-do is for you; wasn't sure how to add it myself.

`.form_label_hidden` sounds good.

Also would like your :+1: on the implementation. It's a little awkward from an i18n perspective, but [this comment](https://github.com/plataformatec/simple_form/issues/25#issuecomment-158633581) gives the impression that there's no better approach... :confused:

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 1]
```diff
@@ -431,24 +436,30 @@ form {
...
   }
 
   .form_item_horiz_input {
-    width: 75%;
+    width: 70%;
   }
 
   .form_actions.horizontal {
-    padding-left: 24.99999%;
+    padding-left: 30%;
```
Comment (ajb) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Can you describe this change?

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 2]
```diff
@@ -408,19 +409,23 @@ form {
   margin-bottom: $rhythm * 2;
 }
 
-.form_item_horiz_label {
+.form_item_horiz_label,
+.form_item_horiz legend {
```
Comment (ajb) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Any reason to not just apply the `form_item_horiz_label` class to the `<legend>` tag?

File: vendor/assets/stylesheets/dvl/core/forms.scss [hunk 3]
```diff
@@ -408,19 +409,23 @@ form {
...
   }
-  .radio_buttons &,
-  .boolean &,
-  .check_boxes & {
-    margin-top: $rhythm / 2; // Vertically centering with .control_styled
-  }
 }
 
-.form_item_vert_label {
+.form_item_horiz legend,
```
Comment (ajb) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Can we revert the syntax to `.radio_buttons &,` for a cleaner diff?

**Comment by ajb** [comment]
Sorry this is still sitting. I want to get to it, but the implementation needs to change pretty much wholesale, in order to work properly with `simple_form`. 

If you can answer my Q's inline about the CSS, that will help me figure out how to properly implement it.

File: vendor/assets/stylesheets/dvl/core/forms.scss (hunk 1 above)
Comment (jrubenoff) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: I increased the line length for horizontal labels, and changed this in kind for proper alignment.

File: vendor/assets/stylesheets/dvl/core/forms.scss (hunk 2 above)
Comment (jrubenoff) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: I wanted to avoid having a single class apply to both a an element's container and the element itself. That gets difficult to maintain quickly. We could wrap every `legend` in a `div.form_item_horiz_label`, but that just seemed laborious to do manually every time.

File: vendor/assets/stylesheets/dvl/core/forms.scss (hunk 1 above)
Comment (ajb) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: I understand. But why the change from 25% to 30%? Are we running out of room somewhere? Just trying to understand how this fits into the fieldset stuff.

File: vendor/assets/stylesheets/dvl/core/forms.scss (hunk 3 above)
Comment (jrubenoff) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Done

File: vendor/assets/stylesheets/dvl/core/forms.scss (hunk 1 above)
Comment (jrubenoff) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: While I was working on this issue and staring at our forms, I just felt that it was more comfortable to read.

**Comment by jrubenoff** [comment]
Alright, I'm fine with whatever implementation you decide as long as the markup and CSS classes remain roughly the same.

If we have to choose one, I am strongly biased towards semantics / accessibility over developer efficiency.

**Comment by ajb** [comment]
> If we have to choose one, I am strongly biased towards semantics / accessibility over developer efficiency.

I would not like to sacrifice either for the other.

File: vendor/assets/stylesheets/dvl/core/forms.scss (hunk 1 above)
Comment (ajb) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: I think I've said this before, but if you make changes like this in a separate commit, it's better because:

a) I can understand your intent without this back-and-forth, and
b) it can be cherry-picked out of the PR.

tl;dr if you care about 25% vs 30%, we could merge it in today _outside_ of the context of this PR.

Comment (jrubenoff) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: I know; I guess I'm wondering why it's worth the cognitive overhead for me to do so when I'm just working on something in context, especially when it's a one-line change.

Comment (ajb) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: You don't have to write a commit message immediately when you make the
change. You can wait until you're ready to commit, and then commit
line-by-line. I use GitHub for Mac to do this easily. Does that help?

On Wed, Mar 23, 2016 at 12:04 PM, Josh Rubenoff <notifications@github.com>
wrote:

> In vendor/assets/stylesheets/dvl/core/forms.scss
> <https://github.com/dobtco/dvl-core/pull/243#discussion_r57186751>:
>
> >    }
> >
> >    .form_actions.horizontal {
> > -    padding-left: 24.99999%;
> > +    padding-left: 30%;
>
> I know; I guess I'm wondering why it's worth the cognitive overhead for me
> to do so when I'm just working on something in context, especially when
> it's a one-line change.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
> <https://github.com/dobtco/dvl-core/pull/243/files/2c0664d134b8e5b0a8bb5e05102b91dd41e07840#r57186751>
>



-- 
Adam Becker
(951) 9-BECKER
@AdamJacobBecker

Comment (jrubenoff) [inline vendor/assets/stylesheets/dvl/core/forms.scss]: Sure

**Comment by ajb** [comment]
Closed by #251.
